src/System/Remote/Monitoring/CloudWatch.hs

```diff
@@ -141,36 +149,56 @@ diffSamples prev !curr = Map.foldlWithKey' combine Map.empty curr
         d2 {Distribution.count = Distribution.count d2 - Distribution.count d1}
     diffMetric _ _ = Nothing
 
-flushSample :: CloudWatchEnv -> Metrics.Sample -> IO ()
-flushSample CloudWatchEnv{..} = void . Map.traverseWithKey flushMetric
+metricToDatum :: [Dimension] -> Text -> Metrics.Value -> Maybe MetricDatum
+metricToDatum dim name val = case val of
+  Metrics.Counter n ->
+    Just $ mkDatum (mdValue ?~ fromIntegral n)
+  Metrics.Gauge n ->
+    Just $ mkDatum (mdValue ?~ fromIntegral n)
+  Metrics.Distribution d ->
+    fmap (\dist -> mkDatum (mdStatisticValues ?~ dist)) (conv d)
+  Metrics.Label l ->
+    Nothing
   where
-    flushMetric :: Text -> Metrics.Value -> IO ()
-    flushMetric name =
-      \case
-        Metrics.Counter n ->
-          sendMetric name (mdValue ?~ fromIntegral n)
-        Metrics.Gauge n ->
-          sendMetric name (mdValue ?~ fromIntegral n)
-        Metrics.Distribution d ->
-          forM_ (conv d) $ \dist ->
-            sendMetric name (mdStatisticValues ?~ dist)
-        Metrics.Label l ->
-          pure ()
-
-    sendMetric :: Text -> (MetricDatum -> MetricDatum) -> IO ()
-    sendMetric name k = do
-      e <- trying _Error . void . runResourceT . runAWS cweAwsEnv . send $
-        putMetricData cweNamespace
-          & pmdMetricData .~
-            [ metricDatum name
-              & mdDimensions .~ cweDimensions
-              & k
-            ]
-      case e of
-        Left err -> cweOnError err
-        Right _ -> pure ()
+    mkDatum k =
+      metricDatum name & mdDimensions .~ dim & k
 
     conv :: Distribution.Stats -> Maybe StatisticSet
     conv Distribution.Stats {..} = do
       guard (count > 0)
       pure (statisticSet (fromIntegral count) sum min max)
+
+weighDatum :: MetricDatum -> Int
+weighDatum = BS.length . toBS . toQueryList "member" . (:[])
+
+data SplitAcc = SplitAcc
+  { splitAccData :: !(NonEmpty [MetricDatum])
+  , splitAccSize :: !Int
+  }
+
+splitAt40KB :: [MetricDatum] -> NonEmpty [MetricDatum]
+splitAt40KB = splitAccData . foldl' go (SplitAcc ([] :| []) 0)
+  where
+    go (SplitAcc (acc :| accs) size) x
+      | size + weight >= 38000 =
```

Comment:
I usually lift out and explicitly name fudge factors.

```
fudgeFactor = 2000
limit = 4000
size + weight >= (limit - fudgeFactor)
```

Reply:
:+1:

Reply:
Also I'm unclear on whether AWS meant 1,024 or 1,000. This is probably worth ascertaining even if we can pessimistically opt for the decimal kilo.